app/upgrades/v18/upgrades.go

```diff
@@ -0,0 +1,216 @@
+		ctx.Logger().Info("Starting upgrade v18...")
+
+		ctx.Logger().Info("Updating redemption rate bounds...")
+		UpdateRedemptionRateBounds(ctx, stakeibcKeeper)
+
+		ctx.Logger().Info("Updating unbonding records...")
+		err := UpdateUnbondingRecords(ctx, stakeibcKeeper, recordsKeeper)
+		if err != nil {
+			return vm, err
+		}
+
+		return mm.RunMigrations(ctx, configurator, vm)
+	}
+}
+
+// Updates the outer redemption rate bounds
+func UpdateRedemptionRateBounds(ctx sdk.Context, k stakeibckeeper.Keeper) {
+	for _, hostZone := range k.GetAllHostZone(ctx) {
+		// Give osmosis a bit more slack since OSMO stakers collect real yield
+		outerAdjustment := RedemptionRateOuterMaxAdjustment
+		if hostZone.ChainId == OsmosisChainId {
+			outerAdjustment = outerAdjustment.Add(OsmosisRedemptionRateBuffer)
+		}
+
+		outerMinDelta := hostZone.RedemptionRate.Mul(RedemptionRateOuterMinAdjustment)
+		outerMaxDelta := hostZone.RedemptionRate.Mul(outerAdjustment)
+
+		outerMin := hostZone.RedemptionRate.Sub(outerMinDelta)
+		outerMax := hostZone.RedemptionRate.Add(outerMaxDelta)
+
+		hostZone.MinRedemptionRate = outerMin
+		hostZone.MaxRedemptionRate = outerMax
+
+		k.SetHostZone(ctx, hostZone)
+	}
+}
+
+// Modify HostZoneUnbonding and UserRedemptionRecords NativeTokenAmount to reflect new data structs
+func UpdateUnbondingRecords(ctx sdk.Context, k stakeibckeeper.Keeper, rk recordskeeper.Keeper) error {
+	for _, epochUnbondingRecord := range rk.GetAllEpochUnbondingRecord(ctx) {
+		// loop over host zone unbonding records
+		for _, hostZoneUnbonding := range epochUnbondingRecord.GetHostZoneUnbondings() {
+			recordStatus := hostZoneUnbonding.Status
+			if recordStatus != recordtypes.HostZoneUnbonding_EXIT_TRANSFER_QUEUE {
+				continue // can ignore this record, as it's not unbonding
+			}
+			amountUnbonded := sdkmath.ZeroInt()
+			// check if the unbond amount is stored in cache above
+			allAmountsUnbondedForHost, found := TotalAmountUnbonded[hostZoneUnbonding.HostZoneId]
+			if !found {
+				// check if this unbond time is in the cache above
+				unbondTime := strconv.FormatUint(hostZoneUnbonding.UnbondingTime, 10)
+				for unbondTimePrefixInCache, amountUnbondedInCache := range allAmountsUnbondedForHost {
+					if strings.HasPrefix(unbondTime, unbondTimePrefixInCache) {
+						amountUnbonded = amountUnbondedInCache
+						break
+					}
+				}
+			}
+			// if we can't find unbonding amount, then we use an estimated redemption rate
+			if amountUnbonded.IsZero() {
+				initialRedemptionRate := InitialRedemptionRates[hostZoneUnbonding.HostZoneId]
+				hostZone, found := k.GetHostZone(ctx, hostZoneUnbonding.HostZoneId)
+				if !found {
+					return errorsmod.Wrapf(stakeibctypes.ErrHostZoneNotFound, "unable to find host zone with chain-id %s", hostZoneUnbonding.HostZoneId)
+				}
+				currentRedemptionRate := hostZone.RedemptionRate
+				blendedRedemptionRate := initialRedemptionRate.Add(currentRedemptionRate).Quo(sdk.NewDecFromInt(sdk.NewInt(2)))
```

Comment:
Just sanity checking — how does order of operations work here? Add should precede Mul, right?

Might be good to add a sanity check below that makes sure blended is between initial and current, just to be safe. Or could add more thorough unit tests.

Reply:
It's a great question! I think Add goes before Quo, although we should test it explicitly with a unit test